feat(docs): guides (#30)
Disploy/disploy: apps/docs/docs/Guide/learning-the-workflow/creating-commands.md

---
sidebar_position: 1
---

# Creating commands

Commands are easy to create with the Disploy Framework, simply export default classes which extend the [`Command` class](/docs/Documentation/framework/classes/Command) and store them in the `commands` directory.

```
├── commands
│   └── ping.ts
```

```ts
// commands/ping.ts
import { Command, type ChatInputInteraction } from '@disploy/framework';

export default class PingCommand extends Command {
	public constructor() {
		super({
			name: 'ping',
			description: 'pong!',
		});
	}

	override async slashRun(interaction: ChatInputInteraction) {
		return void interaction.reply({

> 			content: `hello world!`,

Reviewer: ```suggestion
			content: 'hello world!',
```

- `name` - The name of the command. This is what the user will type to run the command. (e.g. `/ping`)
- `description` - The description of the command. This is what the user will see in the command picker.
- `options` - An array of command options. These are the arguments that the user will provide to the command. (e.g. `/ping test: string`) We're still working on abstracting `options` into our own class, but for now you can use the raw JSON format. See the [Discord docs](https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-option-structure) for more information.

Next, we define the `slashRun` method. This is the method that will be called when the command is run. It takes an interaction of the `CHAT_INPUT` type. We then reply to the interaction with the `hello world!` message.

### `slashRun`

The `slashRun` method is called when the command is run. It takes a single argument, the interaction. The interaction is an object which contains information about the interaction, such as the user who ran the command, the arguments they provided, and the channel the command was run in.

> The `slashRun` method must return a `Promise<void>`, once returned the command will be considered finished; in serverless environments this will kill the process. This allows you to run code after initaliy responding to the interaction. This is useful for doing things like sending followup messages or deferring the interaction to do a long-running task and edit the interaction later.

Reviewer: ```suggestion
The `slashRun` method must return `Promise<void>`, once returned the command will be considered finished; in serverless environments this will kill the process. This allows you to run code after initially responding to the interaction. This is useful for doing things like sending follow-up messages or deferring the interaction to do a long-running task and editing the interaction later.
```

Grammar and spelling

<details>

>   <summary> Click here to see an example of a command with followup messages </summary>

Reviewer: ```suggestion
  <summary>Example of a command with follow-up messages</summary>
```

I'm not too sure if leading and trailing spaces are necessary here, but afaik they're not.

Author: The syntax is very strict here, and you’re forced to by prettier rules and it doesn’t render either.

Reviewer: ```suggestion
  <summary> Example of a command with follow-up messages </summary>
```

```ts
import { Command, type ChatInputInteraction } from '@disploy/framework';

export default class HeyCommand extends Command {
	public constructor() {
		super({
			name: 'hey',
			description: 'heyy!',
		});
	}

	override async slashRun(interaction: ChatInputInteraction) {
		interaction.deferReply(); // Initial response

		await new Promise((resolve) => setTimeout(resolve, 2000));

> 		interaction.editReply({
> 			content: `Just wanted to say hey!`,
> 		}); // Follow up message
> 
> 		// Command is implicitly returned here

Reviewer: ```suggestion
		return void interaction.editReply({
			content: 'Just wanted to say hey!',
		}); // Follow-up message
```

Chances are, the reader already knows how to implicitly return. It might be better to explicitly return here.

Author: True, was just trying to describe to the reader the lifecycle.
